loopchain/baseservice/node_subscriber.py

```diff
@@ -68,60 +87,68 @@ def __del__(self):
                 logging.debug(f"Closing websocket connection to {self._target_uri}...")
                 await websocket.close()
 
-    async def subscribe(self, block_height, event: Event):
-        self._exception = None
-        self._subscribe_event = event
-        await self.close()
+    async def run(self, event, block_height):
+        await self._prepare_connection(event)
+        await self._handshake(block_height)
+        await self.subscribe_loop()
 
+    async def _prepare_connection(self, event):
+        self._subscribe_event = event
+        self._websocket: WebSocketClientProtocol = await websockets.connect(
+            uri=self._target_uri,
+            max_size=4 * conf.MAX_TX_SIZE_IN_BLOCK,
+            loop=MessageQueueService.loop
+        )
+
+    async def _handshake(self, block_height):
+        # TODO: block_height not required in handshake step!
         try:
-            # set websocket payload maxsize to 4MB.
-            self._websocket: WebSocketClientProtocol = await websockets.connect(
-                uri=self._target_uri,
-                max_size=4 * conf.MAX_TX_SIZE_IN_BLOCK,
-                loop=MessageQueueService.loop
-            )
+            await self._subscribe_request(block_height)
+            await self._recv_until_timeout()
+        except Exception as e:
+            logging.debug(f"Exception raised during handshake step: {e}", exc_info=True)
+            await self.close()
+        else:
             logging.debug(f"Websocket connection is completed, with id({id(self._websocket)})")
-            request = Request(
-                method="node_ws_Subscribe",
-                height=block_height,
-                peer_id=ChannelProperty().peer_id
-            )
-            await self._websocket.send(json.dumps(request))
-            await self._subscribe_loop(self._websocket)
+        finally:
+            # set subscribe_event to transit the state to Watch.
+            self._subscribe_event.set()
```

Comment:
if exception is raised during handshake, one must not transit its state to Watch.

Reply:
I see.